src/System.Runtime.Analyzers/Core/CallGCSuppressFinalizeCorrectly.cs

```diff
@@ -27,42 +30,221 @@ public abstract class CallGCSuppressFinalizeCorrectlyAnalyzer : DiagnosticAnalyz
 
         public override void Initialize(AnalysisContext analysisContext)
         {
+            analysisContext.RegisterCompilationStartAction(compilationContext =>
+            {
+                var gcSuppressFinalizeMethodSymbol = compilationContext.Compilation
+                                                        .GetTypeByMetadataName("System.GC")
+                                                        .GetMembers("SuppressFinalize")
+                                                        .OfType<IMethodSymbol>()
+                                                        .SingleOrDefault();
+
+                if (gcSuppressFinalizeMethodSymbol == null)
+                {
+                    return;
+                }
+
+                compilationContext.RegisterOperationBlockStartAction(operationBlockContext =>
+                {
+                    if (operationBlockContext.OwningSymbol.Kind != SymbolKind.Method)
+                    {
+                        return;
+                    }
+
+                    var methodSymbol = (IMethodSymbol)operationBlockContext.OwningSymbol;
+                    if (methodSymbol.IsExtern || methodSymbol.IsAbstract)
+                    {
+                        return;
+                    }
+
+                    var analyzer = new SuppressFinalizeAnalyzer(methodSymbol, gcSuppressFinalizeMethodSymbol, compilationContext.Compilation);
+
+                    operationBlockContext.RegisterOperationAction(analyzer.Analyze, OperationKind.InvocationExpression);
+                    operationBlockContext.RegisterOperationBlockEndAction(analyzer.OperationBlockEndAction);
+                });
+            });
+
+        }
+
+        private class SuppressFinalizeAnalyzer
+        {
+            private enum SuppressFinalizeUsage
+            {
+                CanCall,
+                MustCall,
+                MustNotCall
+            }
+
+            private bool _suppressFinalizeCalled = false;
+
+            private readonly Compilation _compilation;
+            private readonly IMethodSymbol _containingMethodSymbol;
+            private readonly IMethodSymbol _gcSuppressFinalizeMethodSymbol;
+
+            private SemanticModel _semanticModel;
+            private IInvocationExpression _invocationExpression;
+
+            public SuppressFinalizeAnalyzer(IMethodSymbol methodSymbol, IMethodSymbol gcSuppressFinalizeMethodSymbol, Compilation compilation)
+            {
+                this._compilation = compilation;
+                this._containingMethodSymbol = methodSymbol;
+                this._gcSuppressFinalizeMethodSymbol = gcSuppressFinalizeMethodSymbol;
+            }
+
+            public void Analyze(OperationAnalysisContext analysisContext)
+            {
+                _invocationExpression = (IInvocationExpression)analysisContext.Operation;
```

Comment:
ah, i have a vague feeling that the API has changed since I initially wrote this, I remember doing an OperationBlockEnd just to be able to report diagnostics. Will make the change.